keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,255 @@
+
+1. The controller increments the `.status.admission.podSetAssignments.count` and passes the UID of the workload you are folding to the parent workload.
+If we still don’t see the workload being deleted, we at least know it has been counted towards the parent workload and the controller shouldn't fold it again.
+
+## Phases for MVP (alpha)
+
+### Phase 1 - Scale Down
+Scaling down will be the first phase towards MVP because it can be implemented without introducing the Workload Slices. 
+
+Scaling down a RayCluster won’t involve the creation of Workload Slices, instead it’ll involve an update to the current workload, no requeuing.
+
+1. Compare job's PodSet.Count vs Workload.Spec.PodSets[1].Count (worker group) inside the jobframework generic reconciler. 
+2. Call *updateWorkloadToMatchJob()*, this will construct and update the workload and in turn update the PodSet Count field.
+3. Inside the *Update()* method from the *workload_controller* update the workload in cache and in queue. By updating the workload in cache this will update the cluster queue resource usage and by updating the workload in queue this will trigger the scheduler so that it re-assigns the flavors to the already assumed workload and in this way PodSetAssignments will be updated by applying admission based on the new assignments. 
+
+#### Job controller
+
+Rework *equivalentToWorkload()* and *reconcile()* to account for potential differences between the job’s number of replicas and the running workload’s PodSet.Count. 
+
+Given these changes, check if the delta is positive or negative indicating a scaleup/scaledown. If it’s a scaledown the reconciler should trigger an update on the workload to match the new job’s spec and no quota needs to be checked or accounted for, the cluster queue should update the workload resource usage. 
+
+### Phase 2 - Aggregating Workload Slices
+
+In Phase 2, aggregating Workload Slices into the parent workload will be implemented. This doesn't represent a usable feature to end users, but it can be reviewed independently from the phase 3.
+
+### Phase 3 - Scale up with Workload Slices and Scheduling Gates
+
+In Phase 3, scale up will be implemented by introducing Workload Slices and adding Pod scheduling gates as part of Kueue’s mutating admission for RayCluster. 
+
+When the RayCluster scales, the RayCluster webhook would be modified to intercept and "gate" all pods. Every time there’s a resize, you create a dependable (child) workload slice and once it's admitted, you increase the count in the original workload, delete the old workload and remove the schedulingGates. 
+- Pros: You are able to hold the pods added by Kuberay
+- Cons: The fact of having schedulingGates, means we need an API call for every pod, because all pods that are created by the RayCluster are going to have schedulingGates. We need to remove those gates and for every pod you need to make API calls.
```

Comment:
I understand the gates will be added by the RayCluster webhook to the RayCluster worker templates, and removed individually from the Pods by the workload slices controller. That requires the RayCluster controller in KubeRay to play nice and not overriding Kueue changes, right?

Reply:
>  That requires the RayCluster controller in KubeRay to play nice and not overriding Kueue changes, right? 

That's right, and as far as I'm aware, KubeRay does not mutate scheduling gates in any way.

Reply:
FYI @kevin85421 :)